File: astroshield-integration-package/utils/validator.js
/**
 * Client-side validation utility for AstroShield messages
 * 
 * This utility provides methods to validate messages against JSON schemas
 * before sending them to Kafka or the API.
 */
const Ajv = require('ajv');
const addFormats = require('ajv-formats');
const fs = require('fs');
const path = require('path');

// Initialize validator
const ajv = new Ajv({ allErrors: true });
addFormats(ajv);

// Cache for loaded schemas
const schemaCache = {};

/**
 * Load a schema from file
 * @param {string} schemaName - Name of the schema without extension (e.g., 'ss0.sensor.heartbeat')
 * @returns {Object} Loaded JSON schema
 */
function loadSchema(schemaName) {
  if (schemaCache[schemaName]) {
    return schemaCache[schemaName];
  }
  
  try {
    const schemaPath = path.resolve(__dirname, '../schemas', `${schemaName}.schema.json`);
    const schema = JSON.parse(fs.readFileSync(schemaPath, 'utf8'));
    schemaCache[schemaName] = schema;
    return schema;
  } catch (error) {
    throw new Error(`Failed to load schema ${schemaName}: ${error.message}`);
  }
}

/**
 * Validate a message against its schema
 * 
 * @param {Object} message - Message to validate
 * @param {string} schemaName - Schema name without extension
 * @returns {Object} Validation result with valid flag and any errors
 */
function validateMessage(message, schemaName) {
  try {
    const schema = loadSchema(schemaName);
    const validate = ajv.compile(schema);
    const valid = validate(message);
    
    return {
      valid,
      errors: validate.errors || null
    };
  } catch (error) {
    return {
      valid: false,
      errors: [{ message: error.message }]
    };
  }
}

/**
 * Format validation errors into a human-readable string
 * 
 * @param {Array} errors - Validation errors from Ajv
 * @returns {string} Formatted error message
 */
function formatErrors(errors) {
  if (!errors || errors.length === 0) {
    return '';
  }
  
  return errors.map(error => {
    const path = error.instancePath || '';
    const property = error.params.missingProperty 
      ? `missing property '${error.params.missingProperty}'` 
      : (error.params.additionalProperty 
        ? `additional property '${error.params.additionalProperty}'`
        : '');
    
    return `${path} ${error.message} ${property}`.trim();
  }).join('\n');
}

/**
 * Validate a message for a specific topic and format errors if any
 * 
 * @param {Object} message - Message to validate
 * @param {string} topic - Kafka topic (used to determine schema)
 * @returns {Object} Result with isValid flag and formatted error message
 */
function validateMessageForTopic(message, topic) {
  // Map topic to schema name
  const topicToSchema = {
    'ss0.sensor.heartbeat': 'ss0.sensor.heartbeat',
    'ss2.data.state-vector': 'ss2.data.state-vector',
    'ss4.ccdm.detection': 'ss4.ccdm.detection',
    'ss5.launch.prediction': 'ss5.launch.prediction',
    'ss5.telemetry.data': 'ss5.telemetry.data',
    'ss5.conjunction.events': 'ss5.conjunction.events',
    'ss5.cyber.threats': 'ss5.cyber.threats',
    'maneuvers.detected': 'maneuvers.detected'
  };
  
  const schemaName = topicToSchema[topic];
  if (!schemaName) {
    return {
      isValid: false,
      error: `Unknown topic: ${topic}`
    };
  }
  
  const result = validateMessage(message, schemaName);
  
  return {
    isValid: result.valid,
    error: result.valid ? null : formatErrors(result.errors)
  };
}

module.exports = {
  validateMessage,
  validateMessageForTopic,
  formatErrors,
  loadSchema
}; 

File: astroshield-integration-package/utils/validator.py
"""
Client-side validation utility for AstroShield messages

This utility provides methods to validate messages against JSON schemas
before sending them to Kafka or the API.
"""
import os
import json
import jsonschema
from jsonschema import ValidationError
from typing import Dict, Any, Tuple, List, Optional


class SchemaValidator:
    """
    Validator for AstroShield messages using JSON Schema validation
    """
    
    def __init__(self, schema_dir: Optional[str] = None):
        """
        Initialize the validator with schema directory
        
        Args:
            schema_dir: Directory containing JSON schema files. If None, uses ../schemas relative to this file.
        """
        if schema_dir is None:
            schema_dir = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), 'schemas')
        
        self.schema_dir = schema_dir
        self.schema_cache = {}
    
    def load_schema(self, schema_name: str) -> Dict[str, Any]:
        """
        Load a schema from file
        
        Args:
            schema_name: Name of the schema without extension (e.g., 'ss0.sensor.heartbeat')
            
        Returns:
            Loaded JSON schema
            
        Raises:
            FileNotFoundError: If schema file does not exist
            json.JSONDecodeError: If schema is not valid JSON
        """
        if schema_name in self.schema_cache:
            return self.schema_cache[schema_name]
        
        schema_path = os.path.join(self.schema_dir, f"{schema_name}.schema.json")
        
        with open(schema_path, 'r') as f:
            schema = json.load(f)
        
        self.schema_cache[schema_name] = schema
        return schema
    
    def validate_message(self, message: Dict[str, Any], schema_name: str) -> Tuple[bool, Optional[List[str]]]:
        """
        Validate a message against its schema
        
        Args:
            message: Message to validate
            schema_name: Schema name without extension
            
        Returns:
            Tuple of (is_valid, errors) where errors is a list of error messages or None if valid
        """
        try:
            schema = self.load_schema(schema_name)
            jsonschema.validate(instance=message, schema=schema)
            return True, None
        except FileNotFoundError as e:
            return False, [f"Schema not found: {e}"]
        except json.JSONDecodeError as e:
            return False, [f"Invalid schema JSON: {e}"]
        except ValidationError as e:
            return False, [self._format_validation_error(e)]
    
    def validate_message_for_topic(self, message: Dict[str, Any], topic: str) -> Tuple[bool, Optional[str]]:
        """
        Validate a message for a specific topic
        
        Args:
            message: Message to validate
            topic: Kafka topic (used to determine schema)
            
        Returns:
            Tuple of (is_valid, error_message) where error_message is None if valid
        """
        # Map topic to schema name
        topic_to_schema = {
            'ss0.sensor.heartbeat': 'ss0.sensor.heartbeat',
            'ss2.data.state-vector': 'ss2.data.state-vector',
            'ss4.ccdm.detection': 'ss4.ccdm.detection',
            'ss5.launch.prediction': 'ss5.launch.prediction',
            'ss5.telemetry.data': 'ss5.telemetry.data',
            'ss5.conjunction.events': 'ss5.conjunction.events',
            'ss5.cyber.threats': 'ss5.cyber.threats',
            'maneuvers.detected': 'maneuvers.detected'
        }
        
        schema_name = topic_to_schema.get(topic)
        if not schema_name:
            return False, f"Unknown topic: {topic}"
        
        is_valid, errors = self.validate_message(message, schema_name)
        
        if not is_valid and errors:
            return False, "\n".join(errors)
        
        return is_valid, None
    
    def _format_validation_error(self, error: ValidationError) -> str:
        """
        Format a validation error into a human-readable string
        
        Args:
            error: ValidationError from jsonschema
            
        Returns:
            Formatted error message
        """
        path = "/".join(str(p) for p in error.path) if error.path else ""
        path_prefix = f"{path}: " if path else ""
        
        return f"{path_prefix}{error.message}"


# Create a default validator instance for easy import
default_validator = SchemaValidator()

# Module-level functions that use the default validator
def validate_message(message: Dict[str, Any], schema_name: str) -> Tuple[bool, Optional[List[str]]]:
    """
    Validate a message against its schema using the default validator
    
    Args:
        message: Message to validate
        schema_name: Schema name without extension
        
    Returns:
        Tuple of (is_valid, errors) where errors is a list of error messages or None if valid
    """
    return default_validator.validate_message(message, schema_name)

def validate_message_for_topic(message: Dict[str, Any], topic: str) -> Tuple[bool, Optional[str]]:
    """
    Validate a message for a specific topic using the default validator
    
    Args:
        message: Message to validate
        topic: Kafka topic (used to determine schema)
        
    Returns:
        Tuple of (is_valid, error_message) where error_message is None if valid
    """
    return default_validator.validate_message_for_topic(message, topic) 